include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -2523,6 +2789,50 @@ namespace internal
               for (unsigned int q = 0; q < n_q_points; ++q)
                 gradients_quad[(c * dim + d) * n_q_points + q] = tmp_values[q];
             }
+        if (hessians == true)
+          {
+            // diagonal part
+            const unsigned int hdim = (dim * (dim + 1)) / 2;
+            for (unsigned int d = 0; d < dim; ++d)
```

Comment:
Do we need to split this into diagonal and off-diagonal part? Why not simply:
```suggestion
            // diagonal part
            const unsigned int hdim = (dim * (dim + 1)) / 2;
            for (unsigned int d = 0; d < dim + hdim; ++d)
```
and remove the last part?

Reply:
Generally, it might be worth to refactor and modularize this method. The same code is applied on different sub-ranges and different input/output vectors.

Reply:
It is not possible to do it the way you suggest. Also the diagonal and off diagonal part are scaled by different factors. You need to scale the off diagonal entries by to to take the symmetry of the hessians into account.

Reply:
What scaling do you mean?

Reply:
I agree with @peterrum here - all you do here is exchange some entries for each derivative component, so there should be no need to treat the first `dim` components differently from the next `hdim-dim` components. (Note that in the code proposed by @peterrum you only need to run up to `hdim`, not `dim + hdim`, as the former already contains diagonals and off-diagonals.

Reply:
Sorry I was thinking about my splitted for loops in `submit_hessian`. You are right, I changed it.

Reply:
Thanks!